src/librustc_passes/rvalue_promotion.rs

```diff
@@ -104,38 +106,83 @@ struct CheckCrateVisitor<'a, 'tcx: 'a> {
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     in_fn: bool,
     in_static: bool,
-    promotable: bool,
     mut_rvalue_borrows: NodeSet,
     param_env: ty::ParamEnv<'tcx>,
     identity_substs: &'tcx Substs<'tcx>,
     tables: &'a ty::TypeckTables<'tcx>,
     result: ItemLocalSet,
 }
 
+#[must_use]
+#[derive(Debug, PartialEq)]
+enum Promotability {
```

Comment:
Maybe this should be exported and used instead of `bool` elsewhere in the compiler.

Reply:
Are there any other things I'd need to implement (like BitAnd and BitOr) to cover the rest of the compiler?

Reply:
I think `BitOr` should be removed, but otherwise you'd likely just have to try and see.